src/FrameSemantics.cc

```diff
@@ -282,6 +282,666 @@ static Errors resolveModelPoseWithPlacementFrame(
+
+/// \brief Check if type T is an sdf::Frame or sdf::InterfaceFrame
+template <typename T>
+inline constexpr bool IsJoint =
+    std::is_same_v<T, sdf::Joint> || std::is_same_v<T, sdf::InterfaceJoint>;
+
+/////////////////////////////////////////////////
+/// \brief Check if type T is an interface element
+template <typename T>
+inline constexpr bool IsInterfaceElement =
+    std::is_same_v<T, sdf::InterfaceModel> ||
+    std::is_same_v<T, sdf::InterfaceLink> ||
+    std::is_same_v<T, sdf::InterfaceFrame> ||
+    std::is_same_v<T, sdf::InterfaceJoint>;
+
+/////////////////////////////////////////////////
+/// \brief Convert the types sdf::World and sdf::Model to the string literals
+/// "World" and "Model" respectively
+template <typename T>
+inline constexpr const char *elemToString(const T *)
+{
+  // cppcheck-suppress syntaxError
+  // cppcheck-suppress unmatchedSuppression
+  if constexpr (std::is_same_v<T, sdf::World>)
+    return "World";
+  if constexpr (std::is_same_v<T, sdf::Model>)
+    return "Model";
+  return "Unknown";
+}
+
+/////////////////////////////////////////////////
+/// \brief Given the types ElementT and ParentT, determine the FrameType and
+/// string representation of the element.
+template <typename ElementT, typename ParentT>
+constexpr std::pair<sdf::FrameType, const char *>
+getFrameTypeAndRepr(const ParentT *)
+{
+  static_assert(IsModel<ParentT> || std::is_same_v<ParentT, World>,
+                "This function requires ParentT to be either an sdf::World, "
+                "sdf::Model or sdf::InterfaceModel");
+
+  if constexpr (std::is_same_v<ElementT, sdf::Link>)
+  {
+    return {FrameType::LINK, "Link"};
+  }
+  else if constexpr(std::is_same_v<ElementT, sdf::Frame>)
+  {
+    return {FrameType::FRAME, "Frame"};
+  }
+  else if constexpr (std::is_same_v<ElementT, sdf::Joint>)
+  {
+    return {FrameType::JOINT, "Joint"};
+  }
+  if constexpr (std::is_same_v<ElementT, sdf::InterfaceLink>)
+  {
+    return {FrameType::LINK, "InterfaceLink"};
+  }
+  else if constexpr (std::is_same_v<ElementT, sdf::InterfaceJoint>)
+  {
+    return {FrameType::JOINT, "InterfaceJoint"};
+  }
+  else if constexpr (std::is_same_v<ElementT, sdf::InterfaceFrame>)
+  {
+    return {FrameType::FRAME, "InterfaceFrame"};
+  }
+
+  if constexpr (std::is_same_v<ParentT, sdf::World>)
+  {
+    if constexpr(std::is_same_v<ElementT, sdf::Model>)
+    {
+      return {FrameType::MODEL, "Model"};
+    }
+    else if constexpr(std::is_same_v<ElementT, sdf::InterfaceModel>)
+    {
+      return {FrameType::MODEL, "Interface Model"};
+    }
+  }
+  else if constexpr (std::is_same_v<ParentT, sdf::Model>)
+  {
+    if constexpr(std::is_same_v<ElementT, sdf::Model>)
+    {
+      return {FrameType::MODEL, "Nested Model"};
+    }
+    else if constexpr(std::is_same_v<ElementT, sdf::InterfaceModel>)
+    {
+      return {FrameType::MODEL, "Nested Interface Model"};
+    }
+  }
+  else
```

Comment:
I'm not sure if this `else` is too broad or if it should be checking something about the `ParentT`?

Reply:
The only other type would be `InterfaceModel` since the `static_assert` at the beginning of the function ensures that, but I went ahead and added a check for clarity. https://github.com/ignitionrobotics/sdformat/pull/764/commits/f2b2f4793ffcd122d2d3627c20402887c901a0ff